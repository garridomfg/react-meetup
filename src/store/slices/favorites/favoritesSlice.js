import { createSlice } from "@reduxjs/toolkit";

export const favoritesSlice = createSlice({
  name: "favorites",
  initialState: {
    favorites: [],
  },
  reducers: {
    addFavorite: (state, action) => {
      state.favorites = [action.payload, ...state.favorites];
    },
    removeFavorite: (state, action) => {
      state.favorites = action.payload;
    },
  },
});

export const { addFavorite, removeFavorite } = favoritesSlice.actions;
