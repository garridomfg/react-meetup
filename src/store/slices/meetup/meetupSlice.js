import { createSlice } from "@reduxjs/toolkit";

export const meetupSlice = createSlice({
  name: "meetups",
  initialState: {
    meetups: [],
  },
  reducers: {
    setMeetups: (state, action) => {
      state.meetups = action.payload;
    },
    addMeetup: (state, action) => {
      state.meetups = [action.payload, ...state.meetups];
    },
  },
});

export const { setMeetups, addMeetup } = meetupSlice.actions;
