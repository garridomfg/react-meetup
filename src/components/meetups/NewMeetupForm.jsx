import { useDispatch } from "react-redux";
import { addMeetup } from "../../store/slices/meetup/meetupSlice";
import { useForm } from "../../util-hooks/useForm";
import Card from "../ui/Card";
import classes from "./NewMeetupForm.module.css";

export const NewMeetupForm = () => {
  const { formState, onInputChange, onResetForm } = useForm({
    id: undefined,
    title: "",
    image: "",
    address: "",
    description: "",
  });

  const { title, image, address, description } = formState;

  const dispatch = useDispatch();

  const submitHandler = (event) => {
    event.preventDefault();

    if (
      title.trim().length > 1 &&
      image.trim().length > 1 &&
      address.trim().length > 1 &&
      description.trim().length > 1
    ) {
      const meetupToSend = {
        ...formState,
        id: Date.now(),
      };

      dispatch(addMeetup(meetupToSend));
      onResetForm();
    }
  };

  return (
    <Card>
      <form data-testid="form" className={classes.form} onSubmit={submitHandler}>
        <div className={classes.control}>
          <label htmlFor="title">Meetup Title</label>
          <input
            value={title}
            onChange={onInputChange}
            type="text"
            required
            id="title"
            name="title"
            data-testid="title"
          />
        </div>
        <div className={classes.control}>
          <label htmlFor="image">Meetup Image</label>
          <input
            value={image}
            onChange={onInputChange}
            type="url"
            required
            id="image"
            name="image"
            data-testid="image"
          />
        </div>
        <div className={classes.control}>
          <label htmlFor="address">Address</label>
          <input
            value={address}
            onChange={onInputChange}
            type="text"
            required
            id="address"
            name="address"
            data-testid="address"
          />
        </div>
        <div className={classes.control}>
          <label htmlFor="description">Description</label>
          <textarea
            value={description}
            onChange={onInputChange}
            id="description"
            name="description"
            required
            rows="5"
            data-testid="description"
          ></textarea>
        </div>
        <div className={classes.actions}>
          <button type="submit">Add Meetup</button>
        </div>
      </form>
    </Card>
  );
};
