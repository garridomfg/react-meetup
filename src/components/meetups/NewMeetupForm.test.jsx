/* eslint-disable testing-library/no-debugging-utils */
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import { favoritesSlice } from "../../store/slices/favorites/favoritesSlice";
import { meetupSlice } from "../../store/slices/meetup/meetupSlice";
import { screen, render, fireEvent } from "@testing-library/react";
import { NewMeetupForm } from "./NewMeetupForm";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => (fn) => fn(),
}));

const store = configureStore({
  reducer: {
    meetups: meetupSlice.reducer,
    favorites: favoritesSlice.reducer,
  },
});

describe("<NewMeetupForm/>", () => {
  beforeEach(() => jest.clearAllMocks());

  test("Should match with snapshot", () => {
    const view = render(
      <Provider store={store}>
        <NewMeetupForm />
      </Provider>
    );
    expect(view).toMatchSnapshot();
  });

  test("Should change the value from the inputs", () => {
    render(
      <Provider store={store}>
        <NewMeetupForm />
      </Provider>
    );
    const inputTitle = screen.getByTestId("title");
    const inputImage = screen.getByTestId("image");
    const inputAdress = screen.getByTestId("address");
    const inputDescription = screen.getByTestId("description");

    fireEvent.input(inputTitle, { target: { value: "Toscana" } });
    fireEvent.input(inputImage, {
      target: {
        value:
          "https://viajes.nationalgeographic.com.es/medio/2021/05/13/siena_e5a8a829_800x800.jpg",
      },
    });
    fireEvent.input(inputAdress, { target: { value: "Italy" } });
    fireEvent.input(inputDescription, {
      target: { value: "Paisaje de la Toscana" },
    });

    expect(inputTitle.value).toBe("Toscana");
    expect(inputImage.value).toBe(
      "https://viajes.nationalgeographic.com.es/medio/2021/05/13/siena_e5a8a829_800x800.jpg"
    );
    expect(inputAdress.value).toBe("Italy");
    expect(inputDescription.value).toBe("Paisaje de la Toscana");
  });
});
