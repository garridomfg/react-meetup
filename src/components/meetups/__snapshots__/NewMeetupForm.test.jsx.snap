// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewMeetupForm/> Should match with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card"
      >
        <form
          class="form"
          data-testid="form"
        >
          <div
            class="control"
          >
            <label
              for="title"
            >
              Meetup Title
            </label>
            <input
              data-testid="title"
              id="title"
              name="title"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="control"
          >
            <label
              for="image"
            >
              Meetup Image
            </label>
            <input
              data-testid="image"
              id="image"
              name="image"
              required=""
              type="url"
              value=""
            />
          </div>
          <div
            class="control"
          >
            <label
              for="address"
            >
              Address
            </label>
            <input
              data-testid="address"
              id="address"
              name="address"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="control"
          >
            <label
              for="description"
            >
              Description
            </label>
            <textarea
              data-testid="description"
              id="description"
              name="description"
              required=""
              rows="5"
            />
          </div>
          <div
            class="actions"
          >
            <button
              type="submit"
            >
              Add Meetup
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card"
    >
      <form
        class="form"
        data-testid="form"
      >
        <div
          class="control"
        >
          <label
            for="title"
          >
            Meetup Title
          </label>
          <input
            data-testid="title"
            id="title"
            name="title"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="control"
        >
          <label
            for="image"
          >
            Meetup Image
          </label>
          <input
            data-testid="image"
            id="image"
            name="image"
            required=""
            type="url"
            value=""
          />
        </div>
        <div
          class="control"
        >
          <label
            for="address"
          >
            Address
          </label>
          <input
            data-testid="address"
            id="address"
            name="address"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="control"
        >
          <label
            for="description"
          >
            Description
          </label>
          <textarea
            data-testid="description"
            id="description"
            name="description"
            required=""
            rows="5"
          />
        </div>
        <div
          class="actions"
        >
          <button
            type="submit"
          >
            Add Meetup
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
